<!DOCTYPE html>
<html lang="en"><head>
<script src="session-common-steps_files/libs/clipboard/clipboard.min.js"></script>
<script src="session-common-steps_files/libs/quarto-html/tabby.min.js"></script>
<script src="session-common-steps_files/libs/quarto-html/popper.min.js"></script>
<script src="session-common-steps_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="session-common-steps_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="session-common-steps_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="session-common-steps_files/libs/quarto-html/quarto-syntax-highlighting-01c78b5cd655e4cd89133cf59d535862.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.6.36">

  <title>Common data analysis steps in life science projects</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="session-common-steps_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="session-common-steps_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="session-common-steps_files/libs/revealjs/dist/theme/quarto-14d72de84f06f2b03b4ce964173f34e2.css">
  <link href="session-common-steps_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="session-common-steps_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="session-common-steps_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="session-common-steps_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="session-common-steps_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="session-common-steps_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <script src="session-common-steps_files/libs/quarto-diagram/mermaid.min.js"></script>
  <script src="session-common-steps_files/libs/quarto-diagram/mermaid-init.js"></script>
  <link href="session-common-steps_files/libs/quarto-diagram/mermaid.css" rel="stylesheet">
  <script src="session-common-steps_files/libs/kePrint-0.0.1/kePrint.js"></script>
  <link href="session-common-steps_files/libs/lightable-0.0.1/lightable.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Common data analysis steps in life science projects</h1>

<div class="quarto-title-authors">
</div>

</section>
<section id="introduction" class="slide level2">
<h2>Introduction</h2>
<p><br></p>
<p>👋 Welcome to the warm-up session!</p>
<p><br></p>
<div class="fragment">
<p>We will</p>
<div>
<ul>
<li class="fragment">broadly review common data analysis steps in life science projects</li>
<li class="fragment">refresh key concepts that we’ll be building on throughout the week</li>
<li class="fragment">introduce the example dataset</li>
</ul>
</div>
</div>
</section>
<section id="lifecycle-of-data-science" class="slide level2">
<h2>Lifecycle of data science</h2>
<p><em>Simple overview</em></p>
<p><br></p>
<div class="cell" data-reveal="true" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A(Define problem) --&gt; B(Collect data)
  B --&gt; C(Clean &amp; preprocess data)
  C --&gt; E(Explore data)
  E --&gt; F1(Var. Transformations)
  F1 --&gt;F2(Inferential statistics)
  E --&gt; H1(Feature engineering)
  H1 --&gt; H2(Predictive modelling)
  F2 --&gt; I(Communicate results)
  H2 --&gt; I(Communicate results)
</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
</section>
<section id="lifecycle-of-data-science-1" class="slide level2">
<h2>Lifecycle of data science</h2>
<p><em>More realistic case</em></p>
<p><br></p>
<div class="cell" data-reveal="true" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A(Define problem) --&gt; B(Collect data)
  B --&gt; C(Clean &amp; preprocess data)
  C --&gt; E(Explore data)
  E --&gt; F1(Var. Transformations)
  F1 --&gt;F2(Inferential statistics)
  E --&gt; H1(Feature engineering)
  H1 --&gt; H2(Predictive modelling)
  F2 --&gt; I(Communicate results)
  H2 --&gt; I(Communicate results)
  
  I --&gt; A
  F2 --&gt; E       
  F2 --&gt; F1      
  H2 --&gt; E       
  H2 --&gt; H1      
  H1 --&gt; E       
  F1 --&gt; C       
  E --&gt; C        
  C --&gt; B        
  B --&gt; A        
  H1 --&gt; F1      
  F1 --&gt; H1      

</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<div class="fragment">
<p>Looping back examples:</p>
<div>
<ul>
<li class="fragment">Exploration uncovers data issues</li>
<li class="fragment">Data limitations question original design</li>
<li class="fragment">Revisit exploration after inferential analysis</li>
<li class="fragment">Model performance leads back to EDA</li>
<li class="fragment">Feature engineering prompts new transformations</li>
</ul>
</div>
</div>
</section>
<section id="example-data" class="slide level2">
<h2>Example data</h2>
<p><em>Diabetes data to study obesity</em></p>
<p><br></p>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>403 participants were interviewed in a study to understand the prevalence of obesity, diabetes, and other cardiovascular risk factors in central Virginia</li>
<li>The data is available as part of <code>faraway</code> package.</li>
</ul>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<table class="lightable-paper lightable-hover table caption-top" data-quarto-postprocess="true" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; width: auto !important; margin-left: auto; margin-right: auto; font-size: 14px; margin-left: auto; margin-right: auto;">
<thead>
<tr class="header">
<th style="text-align: center;" data-quarto-table-cell-role="th">Abbreviation</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">id</td>
<td style="text-align: center;">Subject ID</td>
</tr>
<tr class="even">
<td style="text-align: center;">chol</td>
<td style="text-align: center;">Total Cholesterol [mg/dL]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">stab.glu</td>
<td style="text-align: center;">Stabilize Glucose [mg/dL]</td>
</tr>
<tr class="even">
<td style="text-align: center;">hdl</td>
<td style="text-align: center;">High Density Lipoprotein [mg/dL]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">ratio</td>
<td style="text-align: center;">Cholesterol / HDL Ratio</td>
</tr>
<tr class="even">
<td style="text-align: center;">glyhb</td>
<td style="text-align: center;">Glycosolated Hemoglobin [%]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">location</td>
<td style="text-align: center;">County: Buckingham or Louisa</td>
</tr>
<tr class="even">
<td style="text-align: center;">age</td>
<td style="text-align: center;">age [years]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">gender</td>
<td style="text-align: center;">gender</td>
</tr>
<tr class="even">
<td style="text-align: center;">height</td>
<td style="text-align: center;">height [in]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">weight</td>
<td style="text-align: center;">weight [lb]</td>
</tr>
<tr class="even">
<td style="text-align: center;">frame</td>
<td style="text-align: center;">frame: small, medium or large</td>
</tr>
<tr class="odd">
<td style="text-align: center;">bp.1s</td>
<td style="text-align: center;">First Systolic Blood Pressure</td>
</tr>
<tr class="even">
<td style="text-align: center;">bp.1d</td>
<td style="text-align: center;">First Diastolic Blood Pressure</td>
</tr>
<tr class="odd">
<td style="text-align: center;">bp.2s</td>
<td style="text-align: center;">Second Systolic Blood Pressure</td>
</tr>
<tr class="even">
<td style="text-align: center;">bp.2d</td>
<td style="text-align: center;">Second Diastolic Blood Pressure</td>
</tr>
<tr class="odd">
<td style="text-align: center;">waist</td>
<td style="text-align: center;">waist [in]</td>
</tr>
<tr class="even">
<td style="text-align: center;">hip</td>
<td style="text-align: center;">hip [in]</td>
</tr>
<tr class="odd">
<td style="text-align: center;">time.ppn</td>
<td style="text-align: center;">Postprandial Time [min] when labs were drawn</td>
</tr>
</tbody>
</table>


</div>
</div>
</div></div>
</section>
<section id="example-data-1" class="slide level2">
<h2>Example data</h2>
<p><em>Diabetes data to study obesity</em></p>
<p><br></p>
<div class="columns">
<div class="column" style="width:45%;">
<ul>
<li>We can calculate BMI as <span class="math inline">\(BMI = 703 \times (weight \; [lb] \; / (height \;[in])^2)\)</span> and define obesity as <span class="math inline">\(BMI \ge 30\)</span> storing this information in <code>obese</code> variable (yes/no).</li>
<li>First few observations are shown on the right.</li>
</ul>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 6
Columns: 21
$ id       &lt;int&gt; 1000, 1001, 1002, 1003, 1005, 1008
$ BMI      &lt;dbl&gt; 22.09, 36.92, 47.95, 18.53, 27.52, 26.39
$ obese    &lt;fct&gt; No, Yes, Yes, No, No, No
$ chol     &lt;int&gt; 203, 165, 228, 78, 249, 248
$ stab.glu &lt;int&gt; 82, 97, 92, 93, 90, 94
$ hdl      &lt;int&gt; 56, 24, 37, 12, 28, 69
$ ratio    &lt;dbl&gt; 3.6, 6.9, 6.2, 6.5, 8.9, 3.6
$ glyhb    &lt;dbl&gt; 4.31, 4.44, 4.64, 4.63, 7.72, 4.81
$ location &lt;fct&gt; Buckingham, Buckingham, Buckingham, Buckingham, Buckingham, B…
$ age      &lt;int&gt; 46, 29, 58, 67, 64, 34
$ gender   &lt;fct&gt; female, female, female, male, male, male
$ height   &lt;dbl&gt; 1.57, 1.63, 1.55, 1.70, 1.73, 1.80
$ weight   &lt;dbl&gt; 54.45, 98.10, 115.20, 53.55, 82.35, 85.50
$ frame    &lt;fct&gt; medium, large, large, large, medium, large
$ bp.1s    &lt;int&gt; 118, 112, 190, 110, 138, 132
$ bp.1d    &lt;int&gt; 59, 68, 92, 50, 80, 86
$ bp.2s    &lt;int&gt; NA, NA, 185, NA, NA, NA
$ bp.2d    &lt;int&gt; NA, NA, 92, NA, NA, NA
$ waist    &lt;dbl&gt; 0.7366, 1.1684, 1.2446, 0.8382, 1.1176, 0.9144
$ hip      &lt;int&gt; 38, 48, 57, 38, 41, 42
$ time.ppn &lt;int&gt; 720, 360, 180, 480, 300, 195</code></pre>
</div>
</div>
</div></div>
</section>
<section id="example-data-2" class="slide level2">
<h2>Example data</h2>
<p><em>Diabetes data to study obesity</em></p>
<p><br></p>
<p>In addition we simulated gene expression for 1000 genes, with some known genes associated with obesity being up-regulated.</p>

<img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-6-1.png" class="quarto-figure quarto-figure-center r-stretch" width="960"></section>
<section>
<section id="data-preparation" class="title-slide slide level1 center">
<h1>Data Preparation</h1>

</section>
<section id="data-preparation-1" class="slide level2 smaller">
<h2>Data Preparation</h2>
<p><br></p>
<div class="cell" data-reveal="true" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A(Define problem) --&gt; B(Collect data)
  B --&gt; C(Clean &amp; preprocess data)
  C --&gt; E(Explore data)
  E --&gt; F1(Var. Transformations)
  F1 --&gt;F2(Inferential statistics)
  E --&gt; H1(Feature engineering)
  H1 --&gt; H2(Predictive modelling)
  F2 --&gt; I(Communicate results)
  H2 --&gt; I(Communicate results)
  
style C fill:#FFCC66,stroke:#333,stroke-width:4px
style F1 fill:#FFCC66,stroke:#333,stroke-width:4px
style H1 fill:#FFCC66,stroke:#333,stroke-width:4px

</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<p><br> <br></p>
<div class="fragment">
<table class="caption-top">
<colgroup>
<col style="width: 21%">
<col style="width: 21%">
<col style="width: 29%">
<col style="width: 27%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Step</th>
<th>What it’s about</th>
<th>What to think about</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>🧼</td>
<td><strong>Data Cleaning</strong></td>
<td>Fixing errors, missing values, and inconsistencies</td>
<td>Is the data valid, complete, and reliable?</td>
</tr>
<tr class="even">
<td>⚙️</td>
<td><strong>Data Pre-processing</strong></td>
<td>Preparing clean data for modeling</td>
<td>Is the data in a model-friendly format?</td>
</tr>
<tr class="odd">
<td>🔢</td>
<td><strong>Var. Transformations</strong></td>
<td>Changing variable scale or structure</td>
<td>Does the variable fit the model assumptions?</td>
</tr>
<tr class="even">
<td>🛠️</td>
<td><strong>Feature Engineering</strong></td>
<td>Creating new informative variables from existing ones</td>
<td>Can we build variables that improve signal and interpretability?</td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="data-preparation-2" class="slide level2">
<h2>Data Preparation</h2>
<p><em>Where Does This Step Belong?</em></p>
<p>🧼 Data Cleaning • ⚙️ Pre-processing • 🔢 Variable Transformation • 🛠️ Feature Engineering</p>
<p><br></p>
<div>
<ul>
<li class="fragment">Convert height from inches to meters <span class="fragment">(🧼 🔢)</span><br>
</li>
<li class="fragment">Encode gender as binary (0 = Female, 1 = Male) <span class="fragment">(⚙️)</span><br>
</li>
<li class="fragment">Create BMI from height and weight <span class="fragment">(🛠️ 🔢)</span><br>
</li>
<li class="fragment">Log-transform (normalize) gene expression values <span class="fragment">(🔢 ⚙️)</span><br>
</li>
<li class="fragment">Remove irrelevant genes with no variation across individuals <span class="fragment">(⚙️ 🧼)</span><br>
</li>
<li class="fragment">Impute missing age values <span class="fragment">(🧼 ⚙️)</span><br>
</li>
<li class="fragment">Create a BMI × gender interaction variable <span class="fragment">(🛠️)</span><br>
</li>
<li class="fragment">Write and apply validation rules: flag heights &lt; 50 or &gt; 90 inches <span class="fragment">(🧼)</span><br>
</li>
<li class="fragment">Detect inconsistent category labels: “male” vs “Male” <span class="fragment">(🧼)</span><br>
</li>
<li class="fragment">Create an average inflammation score from 5 genes <span class="fragment">(🛠)</span></li>
<li class="fragment">Scale gene expression features to mean = 0, SD = 1 <span class="fragment">(⚙️ 🔢)</span><br>
</li>
<li class="fragment">Keep one of a highly correlated features, e.g.&nbsp;keep <code>weight</code> but drop <code>waist</code> <span class="fragment">(⚙️)</span></li>
</ul>
</div>
</section>
<section id="data-preparation-3" class="slide level2 smaller">
<h2>Data Preparation</h2>
<p><em>scaling &amp; normalization</em> <br></p>
<ul>
<li><strong>Scaling</strong> of numerical features
<ul>
<li>Changing the range (scale) of the data to prevent features with larger scales dominating the model.</li>
</ul></li>
<li><strong>Normalization</strong>
<ul>
<li>Changing observations so that they can be described by a normal distribution.</li>
<li>e.g.&nbsp;going from <strong>positive skew</strong>: mode &lt; median &lt; mean</li>
<li>or going from <strong>negative skew</strong>: mode &gt; median &gt; mean</li>
</ul></li>
</ul>

<img data-src="images/skew.png" style="width:80.0%" class="r-stretch"><p><em>Source: https://www.biologyforlife.com/skew.html</em></p>
</section>
<section id="data-preparation-4" class="slide level2">
<h2>Data Preparation</h2>
<p><em>common transformations</em> <br></p>
<p><strong>square-root for moderate skew</strong></p>
<ul>
<li>sqrt(x) for positively skewed data,</li>
<li>sqrt(max(x+1) - x) for negatively skewed data</li>
</ul>
<p><strong>log for greater skew</strong></p>
<ul>
<li>log10(x) for positively skewed data,</li>
<li>log10(max(x+1) - x) for negatively skewed data</li>
</ul>
<p><strong>inverse for severe skew</strong></p>
<ul>
<li>1/x for positively skewed data</li>
<li>1/(max(x+1) - x) for negatively skewed data</li>
</ul>
</section>
<section id="data-preparation-5" class="slide level2">
<h2>Data Preparation</h2>
<p><em>dummy variables</em> <br></p>
<div class="columns">
<div class="column" style="width:40%;">
<ul>
<li>Representing categorical variables with <strong>dummy variables</strong> or <strong>one-hot encoding</strong> to create numerical features.
<ul>
<li>For instance a categorical variable <code>obese</code> with three possible vales (underweight, healthy, overweight) can be transformed into two binary variables: “is_healthy”, and “is_overweight”, where the value of each variable is 1 if the observation belongs to that category and 0 otherwise. Only <span class="math inline">\(k-1\)</span> binary variables to encode <span class="math inline">\(k\)</span> categories.</li>
</ul></li>
<li>In <strong>one-hot encoding</strong> <span class="math inline">\(k\)</span> binary variables are created.</li>
</ul>
</div><div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display margin-caption">
<table class="table caption-top" data-quarto-postprocess="true" style="margin-left: auto; margin-right: auto;">
<caption>Example of obese variable with three categories (underweight/healthy/overweight) encoded as dummy variables</caption>
<thead>
<tr class="header">
<th style="text-align: center;" data-quarto-table-cell-role="th">id</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">obese</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">is_healthy</th>
<th style="text-align: center;" data-quarto-table-cell-role="th">is_overweight</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">902</td>
<td style="text-align: center; border-right: 2px solid black;">Overweight</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: center;">911</td>
<td style="text-align: center; border-right: 2px solid black;">Healthy</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">916</td>
<td style="text-align: center; border-right: 2px solid black;">Healthy</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="even">
<td style="text-align: center;">1171</td>
<td style="text-align: center; border-right: 2px solid black;">Underweight</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">0</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1185</td>
<td style="text-align: center; border-right: 2px solid black;">Healthy</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0</td>
</tr>
</tbody>
</table>


</div>
</div>
</div></div>
</section>
<section id="data-preparation-6" class="slide level2">
<h2>Data Preparation</h2>
<p><em>missing data</em> <br></p>
<ul>
<li><strong>handling missing data</strong> via
<ul>
<li>imputations (mean, median, KNN-based)</li>
<li>deleting strategies such as list-wise deletion (complete-case analysis) or pair-wise deletion (available-case analysis)</li>
<li>choosing algorithms that can handle some extent of missing data, e.g.&nbsp;Random Forest, Naive Bayes</li>
</ul></li>
</ul>
</section>
<section id="data-preparation-7" class="slide level2">
<h2>Data Preparation</h2>
<p><em>Rubin’s (1976) missing data classification system</em> <br></p>
<div class="columns">
<div class="column" style="width:50%;">
<h3 id="mcar">MCAR</h3>
<ul>
<li>missing completely at random</li>
</ul>
<h3 id="mar">MAR</h3>
<ul>
<li>missing at random</li>
<li>two observations for Test 2 deleted where Test 1 <span class="math inline">\(&lt;17\)</span></li>
<li>missing data on a variable is related to some other measured variable in the model, but not to the value of the variable with missing values itself</li>
</ul>
<h3 id="mnar">MNAR</h3>
<ul>
<li>missing not at random</li>
<li>omitting two highest values for Test 2</li>
<li>when the missing values on a variable are related to the values of that variable itself</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="figures/missing-data.jpg"> <span class="citation" data-cites="missing2008"><span>“Missing Data: A Gentle Introduction by Patrick E. McKnight, Katherine M. McKnight, Souraya Sidani, and Aurelio Jose Figueredo”</span> (<a href="#/references" role="doc-biblioref" onclick="">2008</a>)</span></p>
</div></div>
</section>
<section id="data-preparation-8" class="slide level2">
<h2>Data Preparation</h2>
<p><em>handling imbalanced data</em> <br></p>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><strong>handling imbalanced data</strong>
<ul>
<li>down-sampling</li>
<li>up-sampling</li>
<li>generating synthetic instances
<ul>
<li>e.g.&nbsp;with SMOTE <span class="citation" data-cites="fernandez2018smote">(<a href="#/references" role="doc-biblioref" onclick="">Fernández et al. 2018</a>)</span></li>
<li>or ADASYN <span class="citation" data-cites="4633969">(<a href="#/references" role="doc-biblioref" onclick="">He et al. 2008</a>)</span></li>
</ul></li>
</ul></li>
</ul>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-8-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section></section>
<section>
<section id="exploratory-data-analysis-eda" class="title-slide slide level1 center">
<h1>Exploratory Data Analysis (EDA)</h1>

</section>
<section id="exploratory-data-analysis-eda-1" class="slide level2">
<h2>Exploratory Data Analysis (EDA)</h2>
<p><br></p>
<div class="cell" data-reveal="true" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A(Define problem) --&gt; B(Collect data)
  B --&gt; C(Clean &amp; preprocess data)
  C --&gt; E(Explore data)
  E --&gt; F1(Var. Transformations)
  F1 --&gt;F2(Inferential statistics)
  E --&gt; H1(Feature engineering)
  H1 --&gt; H2(Predictive modelling)
  F2 --&gt; I(Communicate results)
  H2 --&gt; I(Communicate results)
  
style E fill:#FFCC66,stroke:#333,stroke-width:4px

</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<p><br></p>
<div class="fragment">
<p>EDA is the process of visually and statistically exploring data to:</p>
<div>
<ul>
<li class="fragment">Understand distributions and variable relationships</li>
<li class="fragment">Detect outliers, missing data, or batch effects</li>
<li class="fragment">Identify patterns and guide modeling decisions</li>
</ul>
</div>
</div>
<div class="fragment">
<p><br></p>
<blockquote>
<p>📌 EDA is essential for choosing the right methods and avoiding incorrect assumptions.</p>
</blockquote>
</div>
</section>
<section id="exploratory-data-analysis-eda-2" class="slide level2 smaller">
<h2>Exploratory Data Analysis (EDA)</h2>
<p><br> <br></p>
<div class="columns">
<div class="column" style="width:35%;">
<p>Using our dataset, we can:</p>
<ul>
<li>📊 Visualize gene expression by obesity status (e.g., histograms, boxplots)</li>
<li>🧮 Explore relationships among clinical variables (e.g., age, BMI, gender)</li>
<li>📉 Detect skewness or low-variance genes</li>
<li>🔎 Use PCA to check for sample clusters or outliers</li>
</ul>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/eda-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section></section>
<section>
<section id="inferential-statistics" class="title-slide slide level1 center">
<h1>Inferential Statistics</h1>

</section>
<section id="inferential-statistics-1" class="slide level2 smaller">
<h2>Inferential Statistics</h2>
<div class="cell" data-reveal="true" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart LR
  A(Define problem) --&gt; B(Collect data)
  B --&gt; C(Clean &amp; preprocess data)
  C --&gt; E(Explore data)
  E --&gt; F1(Var. Transformations)
  F1 --&gt;F2(Inferential statistics)
  E --&gt; H1(Feature engineering)
  H1 --&gt; H2(Predictive modelling)
  F2 --&gt; I(Communicate results)
  H2 --&gt; I(Communicate results)
  
style F2 fill:#FFCC66,stroke:#333,stroke-width:4px

</pre>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<p><br></p>
<div class="fragment">
<p>Inferential statistics allow us to draw conclusions about a population based on a sample, and to test associations between variables.</p>
<p>Common Association Tests:</p>
<div>
<ul>
<li class="fragment">🧮 <strong>t-test / ANOVA</strong>: Compare means between groups (e.g., age in obese vs.&nbsp;non-obese)</li>
<li class="fragment">📈 <strong>Correlation tests</strong>: Assess linear associations between continuous variables (e.g., age vs.&nbsp;gene expression)</li>
<li class="fragment">📊 <strong>Chi-squared / Fisher’s test</strong>: Check association between categorical variables (e.g., gene mutation vs.&nbsp;obesity status)</li>
<li class="fragment">🔍 <strong>Regression models</strong>: Model the relationship between variables while adjusting for covariates (e.g.&nbsp;logistic regression to assess whether gene expression is associated with obesity, controlling for age and gender)</li>
<li class="fragment">🧩 <strong>Mixed-effects models</strong>: Model repeated measurements or hierarchical data while accounting for both fixed effects (e.g., treatment, time) and random effects (e.g., individual variability) e.g., modeling gene expression changes over time in individuals taking a drug vs.&nbsp;controls</li>
<li class="fragment">⏱️ <strong>Survival analysis</strong>: Model time-to-event outcomes and censoring (e.g., time to disease onset in relation to gene expression and clinical predictors)</li>
</ul>
</div>
</div>
<div class="fragment">
<blockquote>
<p>📌 These tests help answer:<br>
<strong>“Is the observed relationship likely to exist beyond our data?”</strong></p>
</blockquote>
</div>
</section>
<section id="inferential-statistics-2" class="slide level2">
<h2>Inferential Statistics</h2>
<p><em>linear expression example I</em></p>
<p><br></p>
<div class="columns">
<div class="column" style="width:45%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ FTO, data = data)

Residuals:
     Min       1Q   Median       3Q      Max 
-13.9652  -4.4319  -0.4238   2.6867  22.6768 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)  28.0064     0.3239  86.475  &lt; 2e-16 ***
FTO           2.2651     0.2868   7.897 2.86e-14 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 6.147 on 395 degrees of freedom
  (6 observations deleted due to missingness)
Multiple R-squared:  0.1363,    Adjusted R-squared:  0.1342 
F-statistic: 62.36 on 1 and 395 DF,  p-value: 2.863e-14</code></pre>
</div>
</div>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<p><br></p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-11-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="inferential-statistics-3" class="slide level2">
<h2>Inferential Statistics</h2>
<p><em>linear expression example II</em></p>
<p><br></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ FTO + age + gender, data = data)

Residuals:
     Min       1Q   Median       3Q      Max 
-12.5894  -3.9624  -0.3899   2.5663  21.6297 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 29.011464   0.940338  30.852  &lt; 2e-16 ***
FTO          2.145884   0.279805   7.669 1.37e-13 ***
age          0.007214   0.018346   0.393    0.694    
gendermale  -3.077730   0.610312  -5.043 7.02e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 5.972 on 393 degrees of freedom
  (6 observations deleted due to missingness)
Multiple R-squared:  0.1888,    Adjusted R-squared:  0.1826 
F-statistic:  30.5 on 3 and 393 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="inferential-statistics-4" class="slide level2">
<h2>Inferential Statistics</h2>
<p><em>logistic regression example I</em></p>
<div class="columns">
<div class="column" style="width:45%;">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = obese ~ FTO, family = binomial, data = data)

Coefficients:
            Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)  -1.0307     0.1489  -6.922 4.45e-12 ***
FTO           1.2236     0.1448   8.447  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 494.5  on 368  degrees of freedom
Residual deviance: 391.4  on 367  degrees of freedom
AIC: 395.4

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<p><br></p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-14-1.png" width="768"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="inferential-statistics-5" class="slide level2">
<h2>Inferential Statistics</h2>
<p><em>logistic regression example II</em></p>
<p><br></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = obese ~ age + gender + FTO, family = binomial, 
    data = data)

Coefficients:
             Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept) -0.963859   0.406969  -2.368 0.017866 *  
age          0.006245   0.008051   0.776 0.437908    
gendermale  -0.869609   0.260870  -3.333 0.000858 ***
FTO          1.205950   0.145611   8.282  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 494.50  on 368  degrees of freedom
Residual deviance: 379.62  on 365  degrees of freedom
AIC: 387.62

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
</section></section>
<section>
<section id="predictive-modelling" class="title-slide slide level1 center">
<h1>Predictive Modelling</h1>

</section>
<section id="predictive-modelling-1" class="slide level2">
<h2>Predictive Modelling</h2>
<p><em>classification and regression</em></p>
<ul>
<li>Focuses on building models that <strong>predict outcomes</strong> based on input features<br>
<em>e.g., predicting obesity status from clinical and gene expression data</em></li>
</ul>
<div class="fragment">
<p><img data-src="images/supervised-02.png" style="width:80.0%;height:80.0%" data-align="center"></p>
</div>
</section>
<section id="predictive-modelling-2" class="slide level2">
<h2>Predictive Modelling</h2>
</section>
<section id="predictive-modelling-3" class="slide level2">
<h2>Predictive Modelling</h2>
<p><em>Many algorithms out there</em></p>
<p><br></p>
<p>Common algorithms:</p>
<ul>
<li>KNN, k-Nearest Neighbors</li>
<li>Random forest</li>
<li>Support vector machines</li>
<li>Regularized models</li>
<li>PLS-based methods</li>
<li>Neural networks</li>
<li>Naive Bayes</li>
<li>Gaussian Processes</li>
<li>Ensemble Methods</li>
</ul>
</section>
<section id="predictive-modelling-4" class="slide level2">
<h2>Predictive Modelling</h2>
<p><em>Many algorithms out there</em></p>
<p><br></p>
<p>Common algorithms:</p>
<ul>
<li>KNN, k-Nearest Neighbors</li>
<li><strong>Random forest</strong></li>
<li><strong>Support vector machines</strong></li>
<li><strong>Regularized models</strong></li>
<li><strong>PLS-based methods</strong></li>
<li><strong>Neural networks</strong></li>
<li>Naive Bayes</li>
<li>Gaussian Processes</li>
<li><strong>Ensemble Methods</strong></li>
</ul>
</section>
<section id="predictive-modelling-5" class="slide level2 smaller">
<h2>Predictive Modelling</h2>
<p><em>Lasso regularization</em></p>
<p><br></p>
<div class="columns">
<div class="column" style="width:45%;">
<p><strong>Linear regression</strong> fits a line by minimizing squared errors</p>
<ul>
<li>⚠️ with many predictors, especially correlated ones,<br>
the model can <strong>overfit</strong> and become <strong>hard to interpret</strong></li>
</ul>
<p><br></p>
<p><strong>Lasso (Least Absolute Shrinkage and Selection Operator)</strong><br>
adds a <strong>penalty on the absolute size</strong> of coefficients:</p>
<p><span class="math display">\[ \text{Minimize: } \text{RSS} + \lambda \sum_{j=1}^{p}|\beta_j|\]</span></p>
<p><br></p>
<p><strong>λ (lambda)</strong> is the regularization parameter:</p>
<ul>
<li>Controls the strength of the penalty<br>
</li>
<li>Large λ → more shrinkage → more coefficients set to <strong>zero</strong> (feature selection)</li>
<li>Small λ → less shrinkage → behaves more like ordinary regression</li>
</ul>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-16-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="session-common-steps_files/figure-revealjs/unnamed-chunk-17-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="data-splitting" class="slide level2">
<h2>Data splitting</h2>
<p><em>data splitting strategies</em></p>
<p><br></p>
<div class="columns">
<div class="column" style="width:45%;">
<p>🧪 <strong>Train / Validation / Test</strong></p>
<p>Split data once:</p>
<ul>
<li>[Train ][Validation ][Test ]</li>
<li>[60–70% data ][15–20% data][15–20% data]</li>
<li>Train model, tune on validation, evaluate on test</li>
</ul>
<p><br></p>
<p>🔁 <strong>k-Fold Cross-Validation (e.g.&nbsp;k = 5)</strong></p>
<p>Rotate test set across folds:</p>
<ul>
<li>[Fold1][Fold2][Fold3][Fold4][Fold5]</li>
<li>Test Train Train Train Train</li>
<li>Every sample is tested once</li>
</ul>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<p>🔁🔁 <strong>Repeated k-Fold Cross-Validation</strong></p>
<ul>
<li>Repeat k-fold multiple times</li>
</ul>
<p><br> <br> <br></p>
<p>🧍 <strong>Leave-One-Out CV (LOOCV)</strong></p>
<p>Each sample is its own test set:</p>
<ul>
<li>[Train: n-1 samples] → [Test: 1 sample]</li>
</ul>
</div></div>
</section>
<section id="model-evaluation-metrics" class="slide level2 smaller">
<h2>Model Evaluation Metrics</h2>
<p><br></p>
<table class="caption-top">
<thead>
<tr class="header">
<th></th>
<th>Predicted: Yes</th>
<th>Predicted: No</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Actual: Yes</strong></td>
<td>True Positive (TP)</td>
<td>False Negative (FN)</td>
</tr>
<tr class="even">
<td><strong>Actual: No</strong></td>
<td>False Positive (FP)</td>
<td>True Negative (TN)</td>
</tr>
</tbody>
</table>
<p><br></p>
<div class="columns">
<div class="column" style="width:45%;">
<p><strong>Classification Metrics</strong></p>
<ul>
<li><strong>Accuracy</strong>
<ul>
<li>Overall proportion of correct predictions<br>
</li>
</ul></li>
<li><strong>Precision</strong> TP / (TP + FP)
<ul>
<li>Of all the samples the model predicted as positive (e.g.&nbsp;obese), how many were actually correct?</li>
</ul></li>
<li><strong>Specificity</strong> TN / (TN + FP)
<ul>
<li>The proportion of actual negatives (e.g.&nbsp;non-obese individuals) that are correctly predicted as negative</li>
</ul></li>
<li><strong>Recall (Sensitivity)</strong> TP / (TP + FN)
<ul>
<li>Of all the actual positives (true obese cases), how many did the model catch?</li>
</ul></li>
<li><strong>F1 Score</strong> Harmonic mean of precision and recall
<ul>
<li>F1 balances both precision and recall</li>
</ul></li>
<li><strong>AUC</strong> – Measures overall discrimination</li>
</ul>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<p><strong>Regression Metrics</strong></p>
<ul>
<li><strong>RMSE</strong> – Root Mean Squared Error<br>
</li>
<li><strong>MAE</strong> – Mean Absolute Error<br>
</li>
<li><strong>R²</strong> – Proportion of variance explained</li>
</ul>
</div></div>
</section>
<section id="bias-ethical-issues-in-the-data-analysis-lifecycle" class="slide level2">
<h2>Bias &amp; Ethical Issues in the Data Analysis Lifecycle</h2>
<p><br></p>
<p>Bias and fairness concerns can appear at <strong>many stages</strong> of a typical data science workflow:</p>
<div>
<ul>
<li class="fragment">🧩 <strong>Problem definition</strong>
<ul>
<li class="fragment">Is the question framed fairly? Who benefits from the results?</li>
</ul></li>
<li class="fragment">📥 <strong>Data collection</strong>
<ul>
<li class="fragment">Underrepresentation of certain groups, Historical or measurement bias</li>
</ul></li>
<li class="fragment">🧼 <strong>Data cleaning / preprocessing</strong>
<ul>
<li class="fragment">Imputation choices may favor dominant groups. Inconsistent category labels can skew analysis</li>
</ul></li>
<li class="fragment">🔍 <strong>Exploration &amp; modeling</strong>
<ul>
<li class="fragment">Spurious associations misinterpreted as causal. Feature selection may reflect biased patterns.</li>
</ul></li>
<li class="fragment">🧠 <strong>Predictive modeling</strong>
<ul>
<li class="fragment">Algorithms may <strong>amplify existing inequalities</strong>. Biased training data → biased predictions</li>
</ul></li>
<li class="fragment">📢 <strong>Communication &amp; use</strong>
<ul>
<li class="fragment">Misleading visualizations or overclaims. Lack of transparency in methods or assumptions</li>
</ul></li>
</ul>
</div>
</section></section>
<section>
<section id="finally" class="title-slide slide level1 center">
<h1>Finally</h1>

</section>
<section id="summary" class="slide level2">
<h2>Summary</h2>
<p><br></p>
<div>
<ul>
<li class="fragment">We reviewed common data analysis steps in life science projects.</li>
<li class="fragment">Even though we all work on different projects and across various types of omics data, many of the core steps in data analysis are shared.</li>
<li class="fragment">Don’t worry if some concepts felt too easy or too advanced — we’ll build and explore them together throughout the week.</li>
</ul>
</div>
<div class="fragment">
<p>And speaking of practice, let’s dive into the tutorial.</p>
</div>
</section>
<section id="thank-you" class="slide level2">
<h2>Thank you</h2>
<p>questions?</p>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-fernandez2018smote" class="csl-entry" role="listitem">
Fernández, Alberto, Salvador Garcia, Francisco Herrera, and Nitesh V Chawla. 2018. <span>“SMOTE for Learning from Imbalanced Data: Progress and Challenges, Marking the 15-Year Anniversary.”</span> <em>Journal of Artificial Intelligence Research</em> 61: 863–905.
</div>
<div id="ref-4633969" class="csl-entry" role="listitem">
He, Haibo, Yang Bai, Edwardo A. Garcia, and Shutao Li. 2008. <span>“ADASYN: Adaptive Synthetic Sampling Approach for Imbalanced Learning.”</span> In <em>2008 IEEE International Joint Conference on Neural Networks (IEEE World Congress on Computational Intelligence)</em>, 1322–28. <a href="https://doi.org/10.1109/IJCNN.2008.4633969">https://doi.org/10.1109/IJCNN.2008.4633969</a>.
</div>
<div id="ref-missing2008" class="csl-entry" role="listitem">
<span>“Missing Data: A Gentle Introduction by Patrick E. McKnight, Katherine M. McKnight, Souraya Sidani, and Aurelio Jose Figueredo.”</span> 2008. <em>Personnel Psychology</em> 61 (1): 218–21. <a href="https://doi.org/10.1111/j.1744-6570.2008.00111_8.x">https://doi.org/10.1111/j.1744-6570.2008.00111_8.x</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="session-common-steps_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="session-common-steps_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="session-common-steps_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="session-common-steps_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 10,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp('/' + window.location.host + '/');
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>